shell/browser/api/electron_api_web_frame_main.cc

```diff
@@ -92,14 +93,15 @@ void WebFrameMain::Destroyed() {
 }
 
 void WebFrameMain::MarkRenderFrameDisposed() {
-  render_frame_ = nullptr;
   render_frame_disposed_ = true;
 }
 
-void WebFrameMain::UpdateRenderFrameHost(content::RenderFrameHost* rfh) {
-  // Should only be called when swapping frames.
-  DCHECK(render_frame_);
-  render_frame_ = rfh;
+content::RenderFrameHost* WebFrameMain::render_frame_host() const {
+  content::WebContents* web_contents =
+      content::WebContents::FromFrameTreeNodeId(frame_tree_node_id_);
+  if (!web_contents)
+    return nullptr;
+  return web_contents->UnsafeFindFrameByFrameTreeNodeId(frame_tree_node_id_);
```

Comment:
Was this strategy chosen because its the only publicly available API to do this?

`content::WebContents::FromFrameTreeNodeId` calls `FrameTreeNode::GloballyFindByID` to perform [a lookup of the ID in a map.](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/web_contents/web_contents_impl.cc;l=613-622;drc=d597d33725f307428aad0a03266e23e5b99fe772)
Then `content::WebContents::UnsafeFindFrameByFrameTreeNodeId` calls `FrameTree::FindByID` to perform [a loop over all FTNs](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/renderer_host/frame_tree.cc;l=202-208;drc=d7a02514bb30afce607817fd2e8ef8a8af559739) which seems odd.

Is it premature optimization to be concerned that each property access of a WebFrameMain requires a loop over all frames in the tree?

Reply:
I looked into how Chrome extensions handle this natively and they use the same `content::WebContents::UnsafeFindFrameByFrameTreeNodeId` method to do so. https://source.chromium.org/chromium/chromium/src/+/main:extensions/browser/extension_api_frame_id_map.cc;l=97-119;drc=8690530ca5d0a06f5708b3891f25aaae308f01f7

The main difference with native Chrome extensions is that their lookups are async. The extension author will request an action to be performed in a specific frame from a background script. During the time between when the request was sent and received over IPC, it's possible for a cross-process navigation to occur which is why it's `Unsafe` in that context.

Their requests are scoped to include a `tabId` and `frameId` which is why it also makes sense to filter by the `WebContents`/tab.

For `WebFrameMain`, our lookups will happen the moment they're requested so it should be safe to lookup the FTN ID directly in a map. No cross-process navigation can occur when a WFM instance is accessed from v8.

Reply:
The benefit of using `content::WebContents::UnsafeFindFrameByFrameTreeNodeId` is that it limits your lookup within the FrameTree of the WebContents. However, the WebContents comes from the initial lookup of the FrameTree from a global ID.

I think we need to skip the WebContents altogether and get the FrameTreeNode directly. It is a private API though which makes it tricky.

Reply:
Yes, I chose this because it is the only publicly available API I could find. My initial thought was to hold a reference to the FrameTreeNode directly, & observe it to monitor deletion, but since that's not a public API, I thought I'd try a different tack. If the public APIs are unsuitable then we should maybe ping someone on Chromium to check and see if we're missing something.